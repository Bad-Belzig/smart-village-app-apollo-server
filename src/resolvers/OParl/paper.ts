import { FilterQuery } from 'mongoose';

import {
  Body,
  Consultation,
  File,
  IPaper,
  IPaperSchema,
  Location,
  Organization,
  Paper,
  Person,
} from '../../models';
import { getPaginatedEntriesByIds } from '../resolverHelpers';

export const paperResolvers = {
  Query: {
    oParlPapers: (
      _,
      args: {
        externalIds?: string[];
        keyword?: string[];
        offset?: number;
        pageSize?: number;
      },
    ) => {
      const filter: FilterQuery<IPaperSchema> = {};

      if (args.keyword?.length) filter.keyword = { $all: args.keyword };

      if (args.externalIds) filter.externalId = { $in: args.externalIds };

      return Paper.find(filter, undefined, {
        limit: args.pageSize,
        skip: args.offset,
      });
    },
  },
  OParlPaper: {
    body: (parent: IPaper) => Body.findOne({ externalId: parent.body }),
    relatedPaper: async (
      parent: IPaper,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Paper,
        parent.relatedPaper,
        args.offset,
        args.pageSize,
      ),
    superordinatedPaper: async (
      parent: IPaper,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Paper,
        parent.superordinatedPaper,
        args.offset,
        args.pageSize,
      ),
    subordinatedPaper: async (
      parent: IPaper,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Paper,
        parent.subordinatedPaper,
        args.offset,
        args.pageSize,
      ),
    mainFile: (parent: IPaper) => File.findOne({ externalId: parent.mainFile }),
    auxiliaryFile: async (
      parent: IPaper,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        File,
        parent.auxiliaryFile,
        args.offset,
        args.pageSize,
      ),
    location: async (
      parent: IPaper,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Location,
        parent.location,
        args.offset,
        args.pageSize,
      ),
    originatorPerson: async (
      parent: IPaper,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Person,
        parent.originatorPerson,
        args.offset,
        args.pageSize,
      ),
    underDirectionOf: async (
      parent: IPaper,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Organization,
        parent.underDirectionOf,
        args.offset,
        args.pageSize,
      ),
    originatorOrganization: async (
      parent: IPaper,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Organization,
        parent.originatorOrganization,
        args.offset,
        args.pageSize,
      ),
    consultation: async (
      parent: IPaper,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Consultation,
        parent.consultation,
        args.offset,
        args.pageSize,
      ),
  },
};
