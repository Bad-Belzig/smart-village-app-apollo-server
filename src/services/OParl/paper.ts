import { fetch } from 'apollo-env';

import { Paper } from '../../models';
import { parsePaper } from '../../parser';

// import flow
// 1. fetch json
// 2. create db object
//   2.1. validation necessary for parsing (done in parser)
//   2.2. parse json
//   2.3. create object
// 3. validate db object
// 4. queue new related objects
// 5. save to db

export const importPaper = async (url: string) => {
  try {
    const response = await fetch(url);

    if (response.ok) {
      const json = await response.json();

      // TODO: check for previous existence

      const lt = new Paper(parsePaper(json));

      await lt.validate();

      // TODO: check for persons, organizations, consultation and files to import

      return lt.save();
    } else {
      throw new Error(
        `Error while fetching Paper from : ${url}! \n response.status: ${response.status}`,
      );
    }
  } catch (e) {
    console.log(e);
  }
};
