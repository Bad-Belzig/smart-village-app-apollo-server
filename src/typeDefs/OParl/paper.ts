import { gql } from 'apollo-server-core';

import { oParlBase } from './oParlBase';

export const paperGQL = gql`
  extend type Query {
    oParlPapers: [OParlPaper!]
  }

  type OParlPaper {
    body: OParlBody
    name: String
    reference: String
    date: Date
    paperType: String
    relatedPaper: [OParlPaper!]
    superordinatedPaper: [OParlPaper!]
    subordinatedPaper: [OParlPaper!]
    mainFile: OParlFile
    auxiliaryFile: [OParlFile!]
    location: [OParlLocation!]
    originatorPerson: [OParlPerson!]
    underDirectionOf: [OParlOrganization!]
    originatorOrganization: [OParlOrganization!]
    consultation: [OParlConsultation!]
    ${oParlBase}
  }
`;
