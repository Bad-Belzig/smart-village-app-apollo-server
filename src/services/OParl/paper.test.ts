import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../helpers';
import { importPaper } from './paper';

const helperFuncs = setupTestDataBase('test_paper_import');

// TODO: change test url to something within our control
const testUrl =
  'https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/papers.asp?id=1198';

describe('importing a paper', () => {
  beforeAll(helperFuncs.beforeAll);

  it('imports properly', basicImportTest(importPaper, testUrl));

  it('import returns undefined on faulty url', invalidUrlTest(importPaper));

  afterAll(helperFuncs.afterAll);
});
