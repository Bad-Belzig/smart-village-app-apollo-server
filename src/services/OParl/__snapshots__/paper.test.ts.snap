// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a paper imports properly for 1.0 1`] = `
Object {
  "__v": 0,
  "auxiliaryFile": Array [],
  "body": "https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/bodies.asp?id=2",
  "consultation": Array [
    "https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/consultations.asp?typ=i&id=9295",
  ],
  "created": 2011-09-28T10:00:00.000Z,
  "date": 2011-10-05T00:00:00.000Z,
  "externalId": "https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/papers.asp?id=1198",
  "keyword": Array [],
  "location": Array [],
  "modified": 2011-10-05T08:30:54.000Z,
  "name": "Beschlussfassung über die Bildung einer Zählkommission",
  "originatorOrganization": Array [],
  "originatorPerson": Array [
    "https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/persons.asp?typ=sb&id=442",
  ],
  "paperType": "Beschlussvorlage öffentlich",
  "reference": "I-2011/034",
  "relatedPaper": Array [],
  "subordinatedPaper": Array [],
  "superordinatedPaper": Array [],
  "type": "https://schema.oparl.org/1.0/Paper",
  "underDirectionOf": Array [
    "https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/organizations.asp?typ=at&id=1",
  ],
  "web": "https://www.lwl-pch.sitzung-online.de/bi/vo020.asp?VOLFDNR=1198",
}
`;

exports[`importing a paper imports properly for 1.0 2`] = `
Object {
  "__v": 0,
  "auxiliaryFile": Array [],
  "body": "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1",
  "consultation": Array [
    "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/consultation/73",
    "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/consultation/72",
  ],
  "created": 2015-09-09T14:01:05.000Z,
  "date": 2015-10-01T00:00:00.000Z,
  "externalId": "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/paper/61",
  "keyword": Array [],
  "location": Array [],
  "mainFile": "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/file/1-1285",
  "modified": 2015-09-09T14:01:05.000Z,
  "name": "Prüfung des Jahresabschlusses zum 31.12.2014 gem. § 101 der Gemeindeordnung",
  "originatorOrganization": Array [],
  "originatorPerson": Array [],
  "paperType": "Beschlussvorlage",
  "reference": "VL-1/2015",
  "relatedPaper": Array [],
  "subordinatedPaper": Array [],
  "superordinatedPaper": Array [],
  "type": "https://schema.oparl.org/1.1/Paper",
  "underDirectionOf": Array [
    "https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/organization/3-2",
  ],
  "web": "https://sdnetrim.kdvz-frechen.de/rim4883/vorgang/?__=UGhVM0hpd2NXNFdFcExjZfP67Id3q2MVIi7RC6r_Q1E",
}
`;
